Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+    THROW_NS_IF_FALSE(E_INVALIDARG, (result != NULL) || (number != NULL));
+
+    NSDecimalCopy(result, number);
+    int exponent;
+    if ((exponent = EXPONENT(result) + power) > EXPONENT_MAX) {
+        INVALIDATE_DECIMAL(result);
+        return NSCalculationOverflow;
+    }
+
+    if (exponent < EXPONENT_MIN) {
+        INVALIDATE_DECIMAL(result);
+        return NSCalculationUnderflow;
+    }
+
+    EXPONENT(result) = exponent;
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Interoperable
+*/
+NSCalculationError NSDecimalNormalize(NSDecimal* number1, NSDecimal* number2, NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, (number1 != NULL) || (number2 != NULL));
+
+    if (NSDecimalIsNotANumber(number1) || NSDecimalIsNotANumber(number2)) {
+        return NSCalculationNoError;
+    }
+
+    if (EXPONENT(number1) == EXPONENT(number2)) {
+        return NSCalculationNoError;
+    }
+
+    // Pick out the larger decimal.
+    NSDecimal* decimalToNormalize;
+    NSDecimal* refDecimal;
+    if (EXPONENT(number1) > EXPONENT(number2)) {
+        decimalToNormalize = number1;
+        refDecimal = number2;
+    } else {
+        decimalToNormalize = number2;
+        refDecimal = number1;
+    }
+
+    // Since we want to keep it whole numbers as per mantissa, we'll normalize the larger of the two
+    int expDiff = EXPONENT(decimalToNormalize) - EXPONENT(refDecimal);
+    for (int index = 0; index < expDiff; index++) {
+        MANTISSA(decimalToNormalize)[LENGTH(decimalToNormalize) + index] = 0;
+    }
+
+    // update the length and the exponent
+    EXPONENT(decimalToNormalize) -= expDiff;
+    LENGTH(decimalToNormalize) += expDiff;
+    return NSCalculationNoError;
+}
+
+/**
+ @Status Stub
+ @Notes
+*/
+NSCalculationError NSDecimalPower(NSDecimal* result, const NSDecimal* number, NSUInteger power, NSRoundingMode roundingMode) {
+    UNIMPLEMENTED();
+    return StubReturn();
+}
+
+static void _roundUp(NSDecimal* decimal) {
+    for (int index = LENGTH(decimal) - 1; index >= 0; index--) {
+        // If it's 9, we set it to 0 and cause a propagation
+        if (MANTISSA(decimal)[index] != 9) {
+            MANTISSA(decimal)[index] += 1;
+            break;
+        }
+        MANTISSA(decimal)[index] = 0;
+    }
+}
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalRound(NSDecimal* result, const NSDecimal* number, NSInteger scale, NSRoundingMode roundingMode) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, (result != NULL) || (number != NULL));
+
+    if (NSDecimalIsNotANumber(number)) {
+        return;
+    }
+    if (scale == NSDecimalNoScale || (LENGTH(number) == 0)) {
+        NSDecimalCopy(result, number);
+        return;
+    }
+
+    int adjustedLengthOfMantissa = EXPONENT(number) + LENGTH(number) + scale;
+    if (adjustedLengthOfMantissa <= 0) {
+        // We pretty much rounded it off fully.
```

Comment:
>// We pretty much rounded it off fully. [](start = 8, length = 39)

Not necessarily. If we round 7 to a scale of -1 with default rounding, it should round up to 10, not down to 0.